magenta/models/polyphonic_rnn/polyphonic_rnn_lib.py

```diff
@@ -0,0 +1,1371 @@
+      notes = ns.notes
+      st = np.array([n.start_time for n in notes]).astype('float32')
+      et = np.array([n.end_time for n in notes]).astype('float32')
+      pi = np.array([n.pitch for n in notes]).astype('float32')
+
+      sample_times = sorted(list(set(st)))
+      # go straight for pitch and delta time encoding
+      sn = self.simultaneous_notes
+      pitch_slices = [pi[st == sti][::-1] for sti in sample_times]
+      # This monster fills in 0s so that array size is consistent
+      pitch_slices = [p[:sn] if len(p) >= sn
+                      else
+                      np.concatenate((p, np.array([0.] * (sn - len(p)),
+                                                  dtype='float32')))
+                      for p in pitch_slices]
+      start_slices = [st[st == sti] for sti in sample_times]
+      end_slices = [et[st == sti] for sti in sample_times]
+      start_slices = [ss[:sn] if len(ss) >= sn
+                      else
+                      np.concatenate((ss, np.array([ss[0]] * (sn - len(ss)),
+                                                   dtype='float32')))
+                      for ss in start_slices]
+      end_slices = [es[:sn] if len(es) >= sn
+                    else
+                    np.concatenate((es, np.array([max(es)] * (sn - len(es)),
+                                                 dtype='float32')))
+                    for es in end_slices]
+      start_slices = np.array(start_slices)
+      end_slices = np.array(end_slices)
+      delta_slices = end_slices - start_slices
+      all_ds.append(np.array(delta_slices))
+      all_ps.append(np.array(pitch_slices))
+    assert len(all_ds) == len(all_ps)
+    if new_file_new_sequence:
+      raise ValueError('Unhandled case')
+    else:
+      if not make_augmentations:
+        all_ds = np.concatenate(all_ds)
+        all_ps = np.concatenate(all_ps)
+      else:
+        new_ps_list = []
+        new_ds_list = []
+        assert len(all_ds) == len(all_ps)
+        for n, (ds, ps) in enumerate(zip(all_ds, all_ps)):
+          new_ps_list.append(ps)
+          new_ds_list.append(ds)
+          # Do +- 5 steps for all 11 offsets
+          for i in range(5):
+            new_up = ps + i
+            # Put silences back
+            new_up[new_up == i] = 0.
+            # Edge case... shouldn't come up in general
+            new_up[new_up > 88] = 88.
+            new_down = ps - i
+            # Put silences back
+            new_down[new_down == -i] = 0.
+            # Edge case... shouldn't come up in general
+            new_down[new_down < 0.] = 1.
+
+            new_ps_list.append(new_up)
+            new_ds_list.append(ds)
+
+            new_ps_list.append(new_down)
+            new_ds_list.append(ds)
+        all_ds = np.concatenate(new_ds_list)
+        all_ps = np.concatenate(new_ps_list)
+
+      self._min_time_data = np.min(all_ds)
+      self._max_time_data = np.max(all_ds)
+      self.time_classes = list(np.unique(all_ds.ravel()))
```

Comment:
What about this as an approach: hard code a set of acceptable durations (possibly the one you already have for bach) and just skip any input that has a duration not in that set. This should work with any kind of input and won't cause problems if training and eval sets have different sets of durations.